torch/csrc/jit/passes/utils/check_alias_annotation.cpp

```diff
@@ -0,0 +1,244 @@
+#include "check_alias_annotation.h"
+
+namespace torch {
+namespace jit {
+namespace {
+
+IValue deepCopy(const IValue& self) {
+  // primitive types can be copied directly
+  if (!self.isPtrType()) {
+    return self;
+  }
+
+  // Tensors need special handling, since copy assignment creates an alias
+  if (self.isTensor()) {
+    return IValue(self.toTensor().clone());
+  }
+  if (self.isTensorList()) {
+    std::vector<at::Tensor> newList;
+    for (const auto& oldTensor : self.toTensorListRef()) {
+      newList.push_back(oldTensor.clone());
+    }
+    return newList;
+  }
+
+  // Lists of ivalues should recursively deep copy their contents
+  if (self.isGenericList()) {
+    std::vector<IValue> newList;
+    for (const auto& value : self.toGenericListRef()) {
+      newList.push_back(deepCopy(value));
+    }
+    return newList;
+  }
+
+  // Regular lists can copy assign
+  if (self.isIntList()) {
+    return IValue(self.toIntListRef());
+  } else if (self.isDoubleList()) {
+    return IValue(self.toDoubleListRef());
+  } else if (self.isBoolList()) {
+    return IValue(self.toBoolListRef());
+  } else if (self.isString()) {
+    return IValue(self.toStringRef());
+  }
+
+  // If in the future we add more reference types that are used in aten ops,
+  // we'll have to add them as cases here.
+  AT_ASSERT(false);
+}
+
+Stack deepCopy(const Stack& stack) {
+  Stack ret;
+  for (const auto& v : stack) {
+    ret.push_back(deepCopy(v));
+  }
+  return ret;
+}
+
+bool deepEquals(const IValue& lhs, const IValue& rhs) {
+  // only check tensors for now
+  if (!lhs.isTensor() || !rhs.isTensor()) {
+    return true;
+  }
```

Comment:
Oh god that's so misleading. I'd assume `deepEquals(1, 2)` to be false (similarly `deepEquals(torch.tensor(1.), 2)`) but the return true! We should throw in that case!

Reply:
yeah this is terrible. I'll put up a PR to clean up this testing stuff.